### PR TITLE
Log warnings when too many leases are active

### PR DESCRIPTION
**Pull request by gobins**
This could be one way of solving #3903 .
If this looks alright, I could add a tunable var to set the threshold.

### CODE REVIEW

**Comment by gobins** [comment]
Thanks for the review @jefferai . I have moved the check into the ExpirationManager emitMetrics function.  I have set it to log every 60 seconds. Let me know if this looks alright.

File: vault/expiration.go [hunk 1]
```diff
@@ -1269,6 +1274,17 @@ func (m *ExpirationManager) emitMetrics() {
...
+	// Check if lease count is greater than the threshold
+	if num > maxLeaseThreshold {
+		if atomic.LoadUint32(&m.leaseCheckCounter) > 59 {
+			m.logger.Warn("expiration: lease count exceeds maximum lease threshold")
+			atomic.StoreUint32(&m.leaseCheckCounter, 0)
+		} else {
+			atomic.AddUint32(&m.leaseCheckCounter, 1)
+		}
+	} else {
+		atomic.StoreUint32(&m.leaseCheckCounter, 0)
```
Comment (jefferai) [inline vault/expiration.go]: This `else` clause is not necessary.

File: vault/expiration.go [hunk 2]
```diff
@@ -1269,6 +1274,17 @@ func (m *ExpirationManager) emitMetrics() {
 	num := len(m.pending)
 	m.pendingLock.RUnlock()
 	metrics.SetGauge([]string{"expire", "num_leases"}, float32(num))
+	// Check if lease count is greater than the threshold
+	if num > maxLeaseThreshold {
+		if atomic.LoadUint32(&m.leaseCheckCounter) > 59 {
+			m.logger.Warn("expiration: lease count exceeds maximum lease threshold")
```
Comment (jefferai) [inline vault/expiration.go]: I think "...exceeds warning lease threshold" is nicer or people might think that they have hit a hard max and might hit issues immediately.

File: vault/expiration.go (hunk 1 above)
Comment (gobins) [reply vault/expiration.go]: There was an edge case where the counter will not get reset if the value was flapping over the value. But, thinking about it now, it is insignificant to warrant the else loop running all the time. I will remove it.

File: vault/expiration.go [hunk 3]
```diff
@@ -91,8 +95,9 @@ func NewExpirationManager(c *Core, view *BarrierView) *ExpirationManager {
 		restoreLocks: locksutil.CreateLocks(),
 		quitCh:       make(chan struct{}),
 
-		coreStateLock: &c.stateLock,
-		quitContext:   c.activeContext,
+		coreStateLock:     &c.stateLock,
+		quitContext:       c.activeContext,
+		leaseCheckCounter: 0,
```
Comment (vishalnayak) [inline vault/expiration.go]: Do we need this to be initialized?

File: vault/expiration.go (hunk 1 above)
Comment (jefferai) [reply vault/expiration.go]: It won't get reset, but I think that's totally fine. It won't display _more often_ than once every 60 seconds, it just may display less if it's flapping.

File: vault/expiration.go (hunk 3 above)
Comment (jefferai) [reply vault/expiration.go]: No, but doesn't hurt, going to merge!

**Comment by jefferai** [comment]
Thanks so much!

**Pull request merged**